feat: Assistants API 연결

jayn2u opened the pull request:
# 설명

## OpenAI Assistants
<img width="1822" alt="Screenshot 2025-04-20 at 14 50 52" src="https://github.com/user-attachments/assets/c1e99883-08f4-4748-922b-663b4b99b4b9" />

- 위 사진에서 사용하고 있는 `Assistants` 라는 모델을 생성해서, 해당 모델에게 질문 쿼리를 전달하는 API를 `FastAPI` 서버에 생성했습니다. 또한 `SpringBoot` 서버에서 요청을 보낼 수 있도록 코드를 설계했습니다.

## API 호출 테스트

### SpringBoot Application
<img width="1822" alt="Screenshot 2025-04-20 at 14 54 34" src="https://github.com/user-attachments/assets/11f07a99-6d37-4f9d-a864-35022002007d" />

### FastAPI Application
<img width="1822" alt="Screenshot 2025-04-20 at 14 54 38" src="https://github.com/user-attachments/assets/b462befd-a9b5-4aeb-bb02-d816588067fb" />

- 스프링부트에서 일괄적으로 질문을 보내보고 반응값을 확인하기 위해서 자동화 코드를 생성했습니다.

```txt
Requesting FastAPI URL: http://localhost:9090/api/v1/assistant/
Request Body: ProxyAssistantRequest[text=여긴 장마당이 따로 없는 것 같은데, 살림거리를 어디 가서 사야 합네까?]
질문: 여긴 장마당이 따로 없는 것 같은데, 살림거리를 어디 가서 사야 합네까?
응답: 아, 장마당 대신에 여기서는 대형마트나 슈퍼마켓에서 살림살이를 많이 사요. 예를 들어 이마트, 롯데마트, 홈플러스 같은 곳들이 있죠.  

아, 그리고 질문하실 때는 “살림거리를 어디 가서 사야 합네까?” 보다는 “살림거리를 어디에서 사야 합니까?”가 더 자연스러워요. ‘합네까’ 대신에 ‘합니까’를 쓰는 게 표준어예요.  

혹시 더 궁금한 거 있으면 언제든지 물어보세요!
Thread ID: thread_JgqDvw5qTJlPznoLeulz6hZP
--------------------------------------------------
Requesting FastAPI URL (Follow-up): http://localhost:9090/api/v1/assistant/?thread_id=thread_JgqDvw5qTJlPznoLeulz6hZP
Request Body (Follow-up): ProxyAssistantRequest[text=방금 질문에 대해 좀 더 자세히 설명해주세요.]
후속 질문: 방금 질문에 대해 좀 더 자세히 설명해주세요.
후속 응답: 네, 장마당이 없거나 많이 없는 지역에서는 대체로 대형마트나 슈퍼마켓에서 필요한 물건을 사요. 대형마트는 규모가 크고, 식료품부터 생활용품, 가전제품까지 다양한 물건을 한곳에서 구입할 수 있어요. 

예를 들어, 이마트, 롯데마트, 홈플러스 같은 곳이 대표적이에요. 이런 곳은 도시 곳곳에 있어서 접근하기도 편하고, 가격도 비교적 합리적이에요. 그리고 할인 행사도 자주 하니까 잘 활용하면 좋아요.  

또한, 요즘은 온라인 쇼핑도 많이 하는데, 쿠팡, G마켓, 11번가 같은 인터넷 쇼핑몰에서 집에서 간편히 주문할 수도 있어요. 배달도 빠르고 편리해서 많이 쓰인답니다.  

참, 질문할 때 “방금 질문에 대해 좀 더 자세히 설명해주세요” 대신에 조금 더 자연스럽게 말하려면, “방금 질문한 내용에 대해 좀 더 자세히 설명해 주시겠어요?”라고 하면 더 친근하게 들려요.  

필요하면 언제든지 말씀해 주세요!
Thread ID: thread_JgqDvw5qTJlPznoLeulz6hZP
==================================================
```

- 위의 로그가 테스트의 결과물입니다. 발급받은 `thread_id=thread_JgqDvw5qTJlPznoLeulz6hZP` 값을 사용해서 대화의 문맥을 파악할 수 있도록 설계가 되어있습니다. 
-

## 앞으로 해야 할 일

- 모바일 앱에서 구현하려고 하는 시나리오와 연동되게 구현을 해야 합니다.
- `thread_id` 값을 데이터베이스에 보관하고 작업할 수 있도록 구현을 해야 할 것 같은데, 저에게는 아직 크게 와닿지가 않아서 어려움을 겪고 있습니다.
- 모바일에서 요청을 보내는데, 이것을 어떻게 구현해야 할 지 모르겠네요.
- 모바일 앱에서 시뮬레이션을 한 번 시작하면, 해당 화면을 빠져나가기 전까지 같은 대화문맥으로 응답할 수 있도록 조정해야 합니다.
- 그리고 모바일 앱에서 나갔다고 들어오면 같은 주제여도 새로운 대화가 되도록 설계를 해야겠죠?
- 이것을 ERD 위에 어떻게 구현을 해야 할 지 모르겠네요..
- 그리고 사용자 별로 대화내역을 모으기 때문에 분명 데이터베이스 연동 작업이 필요할 것 같은데, 그러면 응답받은 텍스트들을 전부 데이터베이스에 보관하고, `thread_id` 역시 보관을 해야 할 것 같습니다.
- 이제 FastAPI 서버는 어느정도 개발이 끝난 것 같아요.
- SpringBoot에서 응답받은 텍스트와 `thread_id`를 재활용할 수 있도록 비즈니스 로직을 구축하는 것이 중요합니다.
- FastAPI는 이제 성능개선에 집중한달까요...? 공공데이터를 활용해서 조금 더 구체적인 대답을 할 수 있도록 조정하는 것과 프롬프트 엔지니어링, 체이닝 등 다양한 방법을 활용해서 성능개선에 집중을 해야 합니다.

jayn2u committed:
feat: STT-ChatGPT-TTS MP3 파일 처리 기능 추가

STT로 변환한 텍스트를 기반으로 ChatGPT 응답을 생성하고 이를 MP3 파일로 반환하는 기능을 추가했습니다. 관련 프로퍼티 및 새로운 API 엔드포인트가 구현되었습니다.

jayn2u committed:
refactor: import 정리 및 코드 스타일 개선

불필요한 순환 및 정렬되지 않은 import들을 정리하고, 코드 가독성을 위해 줄바꿈과 포매팅을 수정했습니다. 이에 따라 코드 유지보수성과 일관성이 향상되었습니다.

jayn2u committed:
feat: 오류 응답 정의 및 API 응답 명세 개선

LlmErrorResponse 클래스를 추가하여 오류 응답 모델을 정의하였고, LlmController의 API 응답 명세에 인증 실패와 관련된 응답 정보를 상세히 명시했습니다. 이를 통해 API 문서화 및 클라이언트와의 명확한 통신이 가능해졌습니다.

jayn2u committed:
feat: STT-ChatGPT-TTS 기능 추가

사용자가 업로드한 오디오 파일로 STT, ChatGPT, TTS를 연결하여 응답을 처리하는 기능을 추가했습니다. 새로운 탭 및 UI 섹션을 추가하고 관련 이벤트 핸들러를 구현했습니다.

jayn2u committed:
refactor: 코드 정리 및 불필요한 import 제거
- 테스트 코드 및 주요 클래스에서 불필요한 import 문 정리
- 코드 포맷팅 변경 및 가독성 개선
- 사용되지 않는 LlmErrorResponse 클래스 제거

jayn2u committed:
refactor: import 정리 및 코드 가독성 개선

jayn2u committed:
feat: Assistant API 기능 추가 및 기존 LLM 컨트롤러 리팩토링

- Assistant 기능 관련 Controller, Service, Request, Response 구현
- 기존 LLM 컨트롤러에서 Assistant API 분리
- application-local.yml 수정: Assistant 관련 경로 및 설정 추가

jayn2u committed:
feat: FastAPI Assistant 직접 호출 테스트 추가

jayn2u committed:
refactor: LlmService 인터페이스 분리 및 구체적인 서비스 인터페이스로 재구성

- LlmService의 역할을 세분화하여 TextToSpeechService, SpeechToTextService, ChatGptService 등으로 분리
- 기존 컨트롤러와 서비스를 새로운 분리된 인터페이스에 맞게 리팩터링
- LlmController, LlmService는 공통 기능 확장을 위해 유보

jayn2u committed:
feat: API 응답 디버깅 로그 추가

jayn2u committed:
refactor: ProxyAssistantRequest 및 ProxyAssistantResponse 구조 변경 및 관련 코드 수정

jayn2u committed:
refactor: Assistant Audio API 응답 처리 방식 변경 (Resource → ByteArrayResource)

jayn2u commented:
동작이 하지 않던 API들을 수정했습니다.
`Assistants` 관련 API를 많이 업데이트 진행했습니다.

jayn2u committed:
refactor: 음성 합성 엔진 provider를 enum 타입으로 변경 및 처리 로직 수정

kwonyj1022 commented on src/test/java/com/newworld/saegil/llm/controller/AssistantControllerDirectApiTest.java:
```diff
@@ -0,0 +1,146 @@
+// Use WebEnvironment.NONE as we are not testing the Spring Boot server itself, just using its context
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.NONE)
+// Rename the class to reflect direct API testing
+public class AssistantControllerDirectApiTest {
```
파이썬 서버를 띄워야만 돌아가는 테스트인지라 이 테스트는 없애거나 주석처리하거나 `@Disabled` 추가해서 무시하도록 해주세요

kwonyj1022 commented on src/test/java/com/newworld/saegil/llm/controller/AssistantControllerDirectApiTest.java:
```diff
@@ -0,0 +1,146 @@
+        ProxyAssistantResponse responseBody = responseEntity.getBody();
+        assertThat(responseBody).isNotNull();
+        // FastAPI might return the original text in a different field or not at all, adjust if needed
+        // assertThat(responseBody.text()).isEqualTo(question);
```
이거 외에도 불필요한 주석은 제거해주세요. 대부분 코드를 보면 뭐 하는 코드인지 알 수 있는 것 같아요

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/service/LlmService.java:
```diff
@@ -1,28 +1,7 @@
-
-    Resource receiveSttChatGptTtsResponseFromAudioFile(MultipartFile multipartFile);
+    // This interface can be removed if no general LLM methods remain,
+    // or kept for future general LLM functionalities.
```
안쓰게 됐다면 그냥 지워도 된다고 생각합니다. 필요하면 그때 가서 필요한 사람이 만들면 되니까요

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/service/ProxyAssistantRequest.java:
```diff
@@ -0,0 +1,10 @@
+public record ProxyAssistantRequest(
+        String text
+) {
+}
```
new line 추가하는 설정이 없어졌나봅니다. 이거 말고도 몇 개 파일에서 경고가 있네요

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/LlmController.java:
```diff
@@ -27,142 +13,7 @@
-    }
+    // LLM 관련 최상위 컨트롤러
+    // 세부 기능은 각 하위 컨트롤러(AssistantController, ChatGptController 등)로 분리되었습니다.
+    // 이 컨트롤러는 향후 공통 LLM 기능이나 라우팅 로직이 필요할 경우 사용될 수 있습니다.
```
이런 주석도 필요 없다고 생각합니다. 리뷰어를 위한 정보라면 pr의 코멘트에 다는 것으로도 충분하다고 생각해요

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/AssistantController.java:
```diff
@@ -0,0 +1,127 @@
+    public ResponseEntity<AssistantResponse> getAssistantResponse(
+            @Parameter(description = "텍스트 쿼리") @RequestBody AssistantRequest request,
+            @Parameter(description = "기존 대화 스레드 ID (선택 사항)") @RequestParam(value = "thread_id", required = false) String threadId
+    ) {
```
```suggestion
    public ResponseEntity<AssistantResponse> getAssistantResponse(
            @Parameter(description = "텍스트 쿼리") 
            @RequestBody AssistantRequest request,
            
            @Parameter(description = "기존 대화 스레드 ID (선택 사항)") 
            @RequestParam(value = "thread_id", required = false) String threadId
    ) {
```

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/AssistantController.java:
```diff
@@ -0,0 +1,127 @@
+    public ResponseEntity<AssistantResponse> getAssistantResponseFromFile(
+            @Parameter(description = "음성 파일 (MP3 등)") @RequestPart("file") MultipartFile multipartFile,
+            @Parameter(description = "기존 대화 스레드 ID (선택 사항)") @RequestParam(value = "thread_id", required = false) String threadId
+    ) {
```
```suggestion
    public ResponseEntity<AssistantResponse> getAssistantResponseFromFile(
            @Parameter(description = "음성 파일 (MP3 등)") 
            @RequestPart("file") MultipartFile multipartFile,
            
            @Parameter(description = "기존 대화 스레드 ID (선택 사항)") 
            @RequestParam(value = "thread_id", required = false) String threadId
    ) {
```

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/AssistantController.java:
```diff
@@ -0,0 +1,127 @@
+            @Parameter(description = "텍스트 쿼리") @RequestBody AssistantRequest request,
+            @Parameter(description = "기존 대화 스레드 ID (선택 사항)") @RequestParam(value = "thread_id", required = false) String threadId,
+            @Parameter(description = "음성 합성 엔진 (openai 또는 elevenlabs, 기본값: openai)") @RequestParam(value = "provider", required = false, defaultValue = "OPENAI") TtsProvider provider
+    ) {
```
```suggestion
    public ResponseEntity<Resource> getAssistantAudioResponse(
            @Parameter(description = "텍스트 쿼리") 
            @RequestBody AssistantRequest request,
            
            @Parameter(description = "기존 대화 스레드 ID (선택 사항)") 
            @RequestParam(value = "thread_id", required = false) String threadId,
            
            @Parameter(description = "음성 합성 엔진 (openai 또는 elevenlabs, 기본값: openai)") 
            @RequestParam(value = "provider", required = false, defaultValue = "OPENAI") TtsProvider provider
    ) {
```

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/AssistantController.java:
```diff
@@ -0,0 +1,127 @@
+                            description = "MP3 파일 생성 성공",
+                            content = @Content(mediaType = MediaType.APPLICATION_OCTET_STREAM_VALUE)
+                    ),
+                    @ApiResponse(
+                            responseCode = "401",
+                            description = "인증 실패 - 로그인이 필요한 기능입니다."
+                    ),
+                    @ApiResponse(
+                            responseCode = "500",
+                            description = "서버 내부 오류"
+                    )
```
정상적인 흐름에 대한 상태코드와 설명은 api마다 다른 정보이기 때문에 필요한 정보이지만, 401과 500은 이 api랑 연관있다기 보다는 공통적인 예외에 해당하기 때문에 굳이 필요 없는 정보라고 생각하는데 이렇게 다 적어주시는 이유가 궁금해요. 저는 모바일 개발을 안해봐서 그럴 순 있지만 도움이 되는 정보인가요?

jayn2u replied:
제가 클라이언트 입장에서는 상태코드에 대한 값에 대한 지식이 없는 사람이라 자세하게 설명을 적어주지 않으면, 문제 원인 파악에 굉장히 오래걸렸어요! 저는 최대한 친절하게 작성하려고 노력합니다. 결국 저희 마음에만 드는 것이 아니라 협업하는데 있어서 드는 비용을 최소화 해아하니까요!

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/AssistantController.java:
```diff
@@ -0,0 +1,127 @@
+
+    @Operation(
+            summary = "MP3 파일로부터 Assistant 응답을 음성으로 가져오기",
+            description = "업로드된 MP3 파일에서 추출한 텍스트를 기반으로 OpenAI Assistant의 응답을 받고 이를 음성으로 변환합니다. `thread_id` 쿼리 파라미터를 포함하면 기존 대화 상태를 유지할 수 있습니다 (선택 사항). `provider` 쿼리 파라미터로 음성 합성 엔진(openai, elevenlabs)을 지정할 수 있습니다.",
```
이런 경우는 `"""`를 사용하면 좋을 것 같아요

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/AssistantController.java:
```diff
@@ -0,0 +1,127 @@
+            @Parameter(description = "음성 파일 (MP3 등)") @RequestPart("file") MultipartFile multipartFile,
+            @Parameter(description = "기존 대화 스레드 ID (선택 사항)") @RequestParam(value = "thread_id", required = false) String threadId,
+            @Parameter(description = "음성 합성 엔진 (openai 또는 elevenlabs, 기본값: openai)") @RequestParam(value = "provider", required = false, defaultValue = "OPENAI") TtsProvider provider
+    ) {
```
```suggestion
    public ResponseEntity<Resource> getAssistantAudioResponseFromFile(
            @Parameter(description = "음성 파일 (MP3 등)") 
            @RequestPart("file") MultipartFile multipartFile,
            
            @Parameter(description = "기존 대화 스레드 ID (선택 사항)") 
            @RequestParam(value = "thread_id", required = false) String threadId,
            
            @Parameter(description = "음성 합성 엔진 (openai 또는 elevenlabs, 기본값: openai)") 
            @RequestParam(value = "provider", required = false, defaultValue = "OPENAI") TtsProvider provider
    ) {
```

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/AssistantController.java:
```diff
@@ -0,0 +1,127 @@
+    ) {
+        log.info("Received Assistant audio file upload request: {}, threadId: {}, provider: {}",
+                multipartFile.getOriginalFilename(), threadId, provider);
+        Resource responseResource = assistantService.getAssistantAudioResponseFromAudioFile(multipartFile, threadId, provider.name().toLowerCase());
```
```suggestion
        Resource responseResource = assistantService.getAssistantAudioResponseFromAudioFile(
                multipartFile, 
                threadId, 
                provider.name().toLowerCase()
        );
```

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/AssistantController.java:
```diff
@@ -0,0 +1,127 @@
+                multipartFile.getOriginalFilename(), threadId, provider);
+        Resource responseResource = assistantService.getAssistantAudioResponseFromAudioFile(multipartFile, threadId, provider.name().toLowerCase());
+        return ResponseEntity.ok()
+                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"assistant_response.mp3\"")
```
이런 파일명은 변동성이 높기 때문에 yml에서 관리하는게 좋을 것 같아요. 파일명 바꿀때마다 새 커밋을 쌓기보다는 배포만 다시 하는게 더 나으니까요

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/AssistantController.java:
```diff
@@ -0,0 +1,127 @@
+@RequestMapping("/api/v1/llm/assistant")
+@RequiredArgsConstructor
+@Tag(name = "Assistant API", description = "OpenAI Assistant API 관련 기능")
+public class AssistantController {
```
final 안 붙은 코드가 많네요

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/controller/AssistantController.java:
```diff
@@ -0,0 +1,127 @@
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.multipart.MultipartFile;
+
+// Enum 정의 추가
+enum TtsProvider {
+    OPENAI, ELEVENLABS
+}
```
별도 클래스로 빼주시고, Service에서도 String을 쓰는 것 대신 이 enum을 사용하는게 좋을 것 같아요

kwonyj1022 reviewed (changes requested):
좀 후루룩 보긴 했는데 리뷰 반영 이후에 좀 더 자세히 살펴볼게요

jayn2u committed:
docs: README 수정 및 Gitignore 업데이트

- README에 프로젝트 개요, 주요 기능, 사용 스택 추가
- Gitignore에 Copilot 관련 파일 추가

jayn2u committed:
refactor: 코드 정리 및 주석 제거, Swagger 응답 추가

kwonyj1022 commented on src/main/java/com/newworld/saegil/llm/config/ProxyProperties.java:
```diff
@@ -1,49 +1,29 @@
+import lombok.Setter;
 import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.stereotype.Component;
+
+@Setter
+@Getter
+@Component
+@ConfigurationProperties(prefix = "llm.proxy")
+public class ProxyProperties {
```
record를 쓰면 좋을 것 같고 `@Component`는 없어도 돼요

jayn2u committed:
refactor: 메서드 파라미터에 final 키워드 추가 및 코드 정리

jayn2u committed:
feat: 음성 파일 응답 파일명 동적 설정을 위한 FileProperties 추가

- FileProperties 추가 및 주입
- ProxyProperties를 record 형태로 변경
- Swagger 문서 설명 개선 (멀티라인 적용)
- TtsProvider를 enum으로 분리
- `assistant_response.mp3` 파일명을 FileProperties 값으로 설정

jayn2u committed:
refactor: LlmService 인터페이스 제거

jayn2u committed:
refactor: 테스트 경로 메서드 변경 및 특정 테스트 제외 설정 추가

jayn2u committed:
feat: 로깅 파일 설정 추가

kwonyj1022 committed:
Merge remote-tracking branch 'origin/develop' into feat/make-assistant-api

# Conflicts:
#	src/main/java/com/newworld/saegil/authentication/controller/AuthenticationController.java
#	src/main/java/com/newworld/saegil/configuration/PropertiesConfiguration.java
#	src/main/java/com/newworld/saegil/exception/GlobalExceptionHandler.java
#	src/main/java/com/newworld/saegil/llm/config/ProxyProperties.java
#	src/main/java/com/newworld/saegil/llm/controller/LlmController.java
#	src/main/java/com/newworld/saegil/llm/service/LlmService.java
#	src/main/java/com/newworld/saegil/notice/domain/Notice.java
#	src/main/java/com/newworld/saegil/notice/service/NoticeService.java
#	src/main/resources/application-local.yml
#	src/test/java/com/newworld/saegil/authentication/service/AuthenticationServiceTest.java

kwonyj1022 committed:
refactor: 파일 끝 개행 추가
